tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/AdoNet/DbScopeFactory.cs

```diff
@@ -0,0 +1,118 @@
+// <copyright file="DbScopeFactory.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System;
+using System.Data;
+using Datadog.Trace.Configuration;
+using Datadog.Trace.ExtensionMethods;
+using Datadog.Trace.Logging;
+
+namespace Datadog.Trace.ClrProfiler.AutoInstrumentation.AdoNet
+{
+    internal static class DbScopeFactory
+    {
+        private static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor(typeof(DbScopeFactory));
+
+        public static Scope CreateDbCommandScope(Tracer tracer, IDbCommand command)
```

Comment:
If I understand correctly, this class is basically an extraction of the previous `ScopeFactory.CreateDbCommandScope`, and shouldn't be used _directly_ by integrations. Integrations should use `DbScopeFactory<T>`, which uses this (along with tests).

To avoid confusion could/should we either move these methods inside `DbScopeFactory<T>` and mark them as `// only public for testing` and/or nest this whole type in `DbScopeFactory` to make it clear that the generic type is the desired entry point?

Reply:
I need to check but I think `DbScopeFactory` shouldn't be a nested class of a generic type. I know if a nested class uses the generic parameter of the parent generic type then the nested class is emitted multiple times, What I don't know is if this applies for nested classes not using the generic parameter type.

As general rule this is good `NormalClass.GenericNestedType<T>`  and not sure on `GenericNestedType<T>.NormalClass`

Reply:
`System.__Cannon` will help with all [reference types](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACAZQAtsoAHAGW2ADoASgFcAdhgCW+GAG4aNANoApcRgDiMETCjiwACgwBPDjAgAzXeLEBKK2gbK1GrTv1GT54uTo2AujWIAzAxkDOqa2mAAwgA2eLgAPAAqAHwMAN40DFkMmdmBwaQMAHIwuBgwACYxcem52dl19cFBnowAEjDR0RAA6tDRFbpWtdRNY8EA7AwARB1dEAx9UAPTsqPjAL6NW9QbQA=)  But is still weird the requirement of use a random `T` when you want to call a method inside the nested class 😄 .

Reply:
`DbScopeFactory<TCommand>` was only used to cache values (abusing static fields in generic types) and should always be used from instrumentation rather than `DbScopeFactory` which doesn't cache anything. To make this clearer and to avoid nesting the non-generic type inside the generic type, I refactored it the other way around. I moved the generic type into the non-generic as `DbScopeFactory.Cache<TCommand>` and made `DbScopeFactory.CreateDbCommandScope()` private since it's only called from the nested class. What do you all think of this approach?